\begin{abstract}
\noindent
Intrinsically disordered regions (IDRs) are segments of proteins without stable three-dimensional structures. As this flexibility allows them to interact with diverse binding partners, IDRs play key roles in cell signaling and gene expression. Despite the prevalence and importance of IDRs in eukaryotic proteomes and various biological processes, associating them with specific molecular functions remains a significant challenge due to their high rates of sequence evolution. However, by comparing the observed values of various IDR-associated properties against those generated under a simulated model of evolution, a recent study found most IDRs across the entire yeast proteome contain conserved features. Furthermore, it showed clusters of IDRs with common ``evolutionary signatures,'' \textit{i.e.} patterns of conserved features, were associated with specific biological functions. To determine if similar patterns of conservation are found in the IDRs of other systems, in this work we apply a series of phylogenetic models to over 8,500 orthologous IDRs identified in the \textit{Drosophila} genome to dissect the forces driving their evolution. By comparing models of constrained and unconstrained continuous trait evolution using the Brownian motion and Ornstein-Uhlenbeck models, respectively, we identify specific clusters of IDRs with shared patterns of constraint. As in yeast, these clusters are enriched for proteins with specific functional annotations, which suggests the preservation of distributed features and their associated functions is a widespread mechanism of IDR evolution.
\end{abstract}

\section{Introduction}
Intrinsically disordered regions (IDRs) are segments of proteins which lack stable three-dimensional structures and instead exist as ensembles of rapidly interconverting conformations. As a result of this structural heterogeneity, IDRs can interact with diverse binding partners. Often these interactions have high specificity but moderate affinity, which permits the efficient propagation of signals by rapid binding and dissociation~\cite{Zhou2012, Wright2014}. Furthermore, as IDRs readily expose their polypeptide chains, they are enriched in recognition motifs for post-translational modifications which allow environmental or physiological conditions to modulate their interactions. Accordingly, IDRs often act as the ``hubs'' of complex signaling networks by integrating signals from diverse pathways and coordinating interactions~\cite{Dunker2005, Buljan2012}. However, as IDRs are pervasive in eukaryotic proteomes, with estimates of the fractions of disordered residues in the human, mouse, and fruit fly proteomes ranging between 22 and 24\%~\cite{Piovesan2020, Piovesan2022}, they are involved in diverse processes~\cite{vanderLee2014} including transcriptional regulation~\cite{Liu2006} and the formation of biomolecular condensates~\cite{Banani2017}.

Despite the prevalence and importance of IDRs in eukaryotic proteomes, associating them with specific molecular functions or biological processes remains a significant challenge. The sequences of IDRs are generally poorly conserved, so traditional bioinformatics approaches which rely on the conservation of amino acid sequences to identify homologous proteins and transfer annotations between them are largely unsuccessful when applied to IDRs. However, several recent studies have demonstrated evidence that IDRs are constrained to preserve ``distributed features'' such as flexibility~\cite{Daughdrill2007}, chemical composition~\cite{Moesa2012}, net charge~\cite{Zarin2017}, or charge distribution~\cite{Beh2012}. Because many sequences can yield a region with a specific composition, for example, this mode of constraint uncouples an IDR's fitness from its strict sequence of amino acid residues. Furthermore, in contrast to folded regions whose precise contacts and packing geometries are easily disrupted by amino acid substitutions, distributed features are robust to such changes, as individual residues only weakly contribute to a region's fitness. For example, a mutation at one site in an IDR that changes its net charge is easily reversed by subsequent compensatory mutations elsewhere in the region. Thus, under this model the sequences of IDRs can rapidly diverge and still preserve their structural or functional properties.

This form of selective constraint can also describe the evolution of more ``localized'' features in IDRs such as short linear motifs (SLiMs). Because SLiMs are composed of fewer than 12 residues, they form limited interfaces that frequently mediate the transient binding events involved in signaling pathways~\cite{Tompa2014}. Accordingly, they are highly enriched in IDRs, which provide an accessible and flexible scaffold for these interactions~\cite{Fuxreiter2007, Davey2012}. While some SLiMs in IDRs are strongly conserved at specific positions, these constitute a small fraction of disordered residues, estimated at roughly 17\% in the yeast proteome~\cite{NguyenBa2012}. Instead, as SLiMs are compact and often highly degenerate at some positions, they can arise de novo from a small number of mutations and therefore have high rates of turnover. Furthermore, when IDRs contain multiple copies of a motif that jointly mediate a high-avidity interaction~\cite{Tompa2014} or a graded response to a signal via the accumulation of multiple phosphorylations~\cite{Wright2014}, the individual motifs are under weak selective constraints. As a result, though SLiMs are encoded by specific sequences, in some contexts they may evolve as distributed features that characterize IDRs as a whole rather than specific sites within them.

The initial studies demonstrating evidence of constraint were generally restricted to specific features or proteins. However, by comparing the observed values of various IDR-associated properties against those generated under a simulated model of evolution, Zarin \textit{et al.} showed most IDRs across the entire yeast proteome contain conserved features. Furthermore, they identified clusters of IDRs with common ``evolutionary signatures,'' \textit{i.e.} patterns of conserved features, which were associated with specific biological functions. This analysis for the first time provided a global view of the relationship between sequence and function in IDRs. A follow-up study then expanded on this initial finding by applying techniques from machine learning and statistics to predict the functions of individual IDRs using their evolutionary signatures~\cite{Zarin2021}.

However, no known subsequent studies have determined if similar patterns of conservation are found in the IDRs of other systems. As another foundational model organism with abundant genomic information across many evolutionary lineages~\cite{Yang2018, Miller2018, Kim2021, Gramates2022}, the fruit fly, \textit{Drosophila melanogaster}, is a natural choice for subsequent investigation. Furthermore, given its complex multicellular development process and shared signaling pathways with humans, the findings of such a study would significantly advance our understanding of the role of IDRs in gene regulation as well as human health and disease. The concordance of these results with the previously identified IDR clusters would also have profound implications for the broader mechanisms of IDR evolution. For example, the absence of global patterns of evolutionary signatures across IDRs in \textit{Drosophila} would suggest they are property of IDRs which is unique to yeast. In contrast, the identification of clusters similar to those in yeast would indicate the existence of a taxonomy of IDRs which is conserved across the tree of life. The latter result would represent a significant step towards the creation of resources for the classification of IDRs analogous to those for folded domains such as Pfam~\cite{Mistry2020}, CATH~\cite{Sillitoe2020}, or SCOP~\cite{Andreeva2013, Andreeva2019}.

Therefore, in this work we apply a series of phylogenetic models to a set of orthologous IDRs identified in the \textit{Drosophila} genome to dissect the forces driving their evolution. Our analyses span multiple levels, ranging from the sequences that compose these regions to the distributed features that characterize them as a whole. For the latter, though the previous approach relied on simulations to generate the null distribution for a hypothesis of no constraint, we instead leverage a fully statistical phylogenetic comparative framework~\cite{Cornwell2017}. By comparing models of constrained and unconstrained continuous trait evolution, \textit{i.e.} the Brownian motion and Ornstein-Uhlenbeck models, respectively, we can demonstrate evidence of selective constraint on features independent of any assumptions about the underlying process of sequence evolution. However, we also explore hybrid approaches that combine simulations with phylogenetic comparative methods to test increasingly refined models of IDR evolution. We find that IDRs exhibit unique patterns of amino acid substitution and that in some proteins disorder itself is a dynamically evolving property. Furthermore, though IDRs are broadly unconstrained along several axes of feature evolution, we identify specific clusters of IDRs with shared patterns of constraint. As in yeast, these clusters are enriched for proteins with specific functional annotations, which suggests the preservation of distributed features and their associated functions is a widespread mechanism of IDR evolution.

\section{Results}
\subsection{IDRs are shorter and more divergent than non-disordered regions}
We identified regions with high levels of inferred intrinsic disorder in over 8,500 alignments of single copy orthologs from 33 species in the \textit{Drosophila} genus using the disorder predictor AUCPreD~\cite{Wang2016}. To highlight the unique features of IDR evolution in subsequent analyses, we also extracted a complementary set of regions with low levels of inferred disorder. Both sets were filtered on several criteria, including the lengths of their sequences and their phylogenetic diversity, which yielded 11,445 and 14,927 regions, respectively, from 8,466 unique alignments. In the subsequent discussion, we refer to these sets as the ``disorder'' and ``order'' regions, respectively. To investigate the broad differences in basic sequence statistics between the two region sets, we first generated histograms from the average length of each region (Fig.~\ref{sfig:regions}). Although both distributions span several orders of magnitude, the order regions are generally longer than the disorder regions, with means of 105 and 245 residues, respectively. We then quantified the sequence divergence in each region by fitting phylogenetic trees to the alignments using amino acid and indel substitution models. The average rates of substitution are significantly larger in the disorder regions, demonstrating that while both sets contain conserved and divergent regions, IDRs are enriched in more rapidly evolving sequences (Fig.~\ref{sfig:regions}).

\subsection{IDRs have distinct patterns of residue substitution}
To gain insight into the substitution patterns of amino acid residues in disorder and order regions, we fit substitution models to meta-alignments sampled from the respective sets of filtered regions. These models are probabilistic descriptions of sequence evolution and are parameterized in terms of the one-way rates of change from one residue to another. Thus, the rates are not necessarily equal for a given pair of residues when the initial and target residues are swapped. For example, the rate of change of valine to tryptophan can be distinct from that of tryptophan to valine. In practice, substitution models are typically constrained to fulfill a condition called time-reversibility, as this converts a difficult multivariate optimization of the tree's branch lengths into a series of simpler univariate optimizations~\cite{Felsenstein1981}. A common method for fulfilling this condition is parameterizing the model in terms of a frequency vector, $\boldsymbol \pi$, and an exchangeability matrix, $S$. The frequency vector determines the model's expected residue frequencies at equilibrium, meaning the model dictates that all sequences eventually approach this distribution, no matter their initial composition. The exchangeability matrix is symmetric ($s_{ij} = s_{ji}$) and encodes the propensity for two residues to interconvert. Because the rate of change from residue $i$ to residue $j$ is given by $r_{ij} = s_{ij}\pi_{j}$, higher exchangeability coefficients yield higher rates of conversion. Thus, exchangeability coefficients are frequently interpreted as a measure of biochemical similarity between residues.

\begin{figure}[h!]
\includegraphics[width=\textwidth]{substitution/out/substitution.png}
\centering
\caption{\textbf{Amino acid substitution models fit to disorder and order regions.}
\textbf{(A)} Amino acid frequencies of substitution models. Amino acid symbols are ordered by their enrichment in disorder regions, calculated as the disorder-to-order ratio of their frequencies. Error bars represent standard deviations over models fit to different meta-alignments ($n = 25$). \textbf{(B-C)} Exchangeability coefficients of disorder and order regions, respectively, averaged over meta-alignments. \textbf{(D)} $\log_{10}$ disorder-to-order ratios of exchangeability coefficients. \textbf{(E-F)} Rate coefficients of disorder and order regions, respectively, averaged over meta-alignments. The vertical and horizontal axes indicate the initial and target amino acids, respectively. \textbf{(G)} $\log_{10}$ disorder-to-order ratios of rate coefficients.}
\label{fig:substitution}
\end{figure}

To highlight the differences in patterns of residue substitution between the disorder and order regions, the parameters in each model are directly compared in Fig.~\ref{fig:substitution}, beginning with the frequency vectors. The disorder regions show an enrichment of ``disorder-promoting'' residues such as serine, proline, and alanine, and a depletion of hydrophobic and bulky residues such as trytophan and phenylalanine (Fig.~\ref{fig:substitution}A). The exchangeability matrices fit to the disorder and order regions have similar overall patterns of high and low coefficients (Fig.~\ref{fig:substitution}B-C). However, the log ratios of the disorder to the order exchangeability coefficients show clear differences within and between the disorder-enriched and -depleted residues. In disorder regions, disorder-enriched residues are less exchangeable with each other, whereas disorder-depleted residues are more exchangeable with each other and with disorder-enriched residues (Fig.~\ref{fig:substitution}D). Likewise, we observe a trend in the log ratios of the rate coefficients where the coefficients above the diagonal are generally positive and those below the diagonal are generally negative. As the coefficients model the one-way rate of substitutions between residues with the vertical and horizontal axes indicating the initial and target residues, respectively, this suggests a net flux towards a more disorder-like composition.

\subsection{Intrinsic disorder is poorly conserved in some proteins}
Though the substitution models reveal specific patterns of evolution at the level of individual residues, the large amounts of sequence divergence between many orthologous IDRs implies their evolution is not well-described by fine-scale models of residue substitution. Given the growing evidence that IDRs are constrained to conserve distributed properties, we instead turned towards characterizing their evolution in terms of 82 disorder-associated ``molecular features'' obtained from the previous study of IDRs in the yeast proteome. However, before conducting an in-depth analysis of these features, we examined the disorder score traces in greater detail and were struck by the significant variability between species. For each residue in the input sequence, AUCPreD returns a score between 0 and 1 where higher values indicate higher confidence in a prediction of intrinsic disorder. In some alignments, the scores would vary by nearly this entire range at a given alignment position even when there was a relatively high level of sequence identity (Fig.~\ref{fig:scores}A).

\begin{figure}[h!]
\includegraphics[width=\textwidth]{scores/out/scores_07E3.png}
\centering
\caption{\textbf{Analyses of disorder scores.}
\textbf{(A)} Example region in the alignment of the sequences in orthologous group 07E3 with their corresponding disorder scores. Higher scores indicate a higher probability of disorder. Disorder score traces are colored by the position of its associated species on the phylogenetic tree. \textbf{(B)} Correlations between disorder scores and feature contrasts in regions. Asterisks indicate statistically significant correlations as computed by permutation tests ($p < 0.001$). \textbf{(C-D)} Example scatter plots showing correlations plotted in panel B. \textbf{(E)} GO term analysis of regions with rapidly evolving disorder scores. Only terms where $p < 0.001$ are shown.}
\label{fig:scores}
\end{figure}

To better understand the relationship of this variability to differences in the regions' biophysical properties, we sought to correlate the average disorder score of the segments in a region with their molecular features. However, as the sequences are not independent but instead related by a hierarchical structure which reflects their evolutionary relationships, any derived features are unsuitable for direct use in many standard statistical procedures. In the most severe cases, traits derived from clades of closely-related species can effectively act as duplicate observations, which can yield spurious correlations. We therefore applied the method of contrasts to both the scores and the features. This algorithm takes differences between adjacent nodes in the phylogenetic tree relating the species to generate ``contrasts,'' which, under some general assumptions of the underlying evolutionary process, are independent and identically-distributed and therefore appropriate for use in correlation analyses. The resulting feature contrasts have varying degrees of correlation with the score contrasts (Fig.~\ref{fig:scores}B-C). Some, like isopoint, are uncorrelated, but most are significantly, if weakly, correlated. In general, the strongest correlations are observed for features which have a direct biophysical relationship to the presence or absence of disorder, such as disorder\_fraction or hydrophobicity. Interestingly, the correlations with many motifs were statistically significant, though small in magnitude relative to the non-motif features. However, a more detailed analysis of this observation is presented in the discussion. To determine if regions with rapidly evolving disorder scores were associated with a particular function, process, or compartment, we then extracted the upper decile of their rate distribution and performed term enrichment analyses on their associated annotations (Fig.~\ref{sfig:scores_histogram}). The most significant terms are generally related to DNA repair or extracellular structure, which suggests these processes and components are enriched in proteins whose structural state is rapidly evolving (Fig.~\ref{fig:scores}E).

\subsection{IDRs have three axes of unconstrained variation}
Having calculated the features associated with the sequence segments composing each region in our data set, we then sought to determine if their distributions contained any global structure which would enable us to identify classes with distinct biophysical or functional properties. These distributions are generated by a complex underlying evolutionary process which reflects the combined effects of selection, drift, and mutation. However, to leverage a statistical framework to infer the properties of this process, we fit Brownian motion (BM) models to the features calculated from the segments in each region. BM is a simple model of evolution where continuous traits change through a series of small, undirected steps. Thus, the traits accumulate variation at a constant rate over time but do not on average deviate from their original values. BM models are therefore specified by two parameters: a rate, which describes the speed at which trait variation accumulates, and a root, which describes the ancestral trait value.

We then applied principal components analyses (PCAs) to visualize the major axes of variation of the root and rate parameters for each feature and region. A difficulty with a direct analysis of the parameter estimates, however, is PCAs are sensitive to differences in scaling between variables, and some features have dramatically different intrinsic scales. For example, many compositional features, like fraction\_S, are restricted to the interval [0, 1] whereas SCD is unbounded and can vary from negative to positive infinity. As a result, SCD is responsible for a significant fraction of the overall variance in both parameter distributions (Fig.~\ref{sfig:root_variance}-\ref{sfig:rate_variance}). Therefore, we first normalized the parameters associated with each feature by transforming them into z-scores relative to their proteome-wide distributions.

\begin{figure}[h!]
\includegraphics[width=\textwidth]{brownian/out/root.png}
\centering
\caption{\textbf{PCA of disorder regions' feature roots.}
\textbf{(A)} The first two PCs of the disorder regions' feature root distributions. The explained variance percentage of each component is indicated in parentheses in the axis labels. \textbf{(B)} The same plot as panel A with the projections of original features onto the components shown as arrows. Only the 16 features with the largest projections are shown. Scaling of the arrows is arbitrary.}
\label{fig:root}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=\textwidth]{brownian/out/rate.png}
\centering
\caption{\textbf{PCA of disorder regions' feature rates.}
\textbf{(A)} The first two PCs of the disorder regions' feature rate distributions. The explained variance percentage of each component is indicated in parentheses. \textbf{(B)} Scatter plot of the disorder regions' feature rates along the first PC against the sum of the average amino acid and indel substitution rates in those regions. \textbf{(C)} The second and third PCs of the disorder regions' feature rate distributions. The explained variance percentage of each component is indicated in parentheses in the axis labels. \textbf{(D)} The same plot as panel C with the projections of original rates onto the components shown as arrows. Only the 16 features with the largest projections are shown. Scaling of the arrows is arbitrary. \textbf{(E-F)} Example alignments of disorder regions from the orthologous groups 0A8A, 3139, 04B0, respectively. The colored bars on the left indicate the hexbin containing that region in panel C.}
\label{fig:rate}
\end{figure}

The first two principal components of the root distributions show little overall structure, though there is a slight enrichment of regions along two axes that correlate with acidic and polar features, respectively (Fig.~\ref{fig:root}). Likewise, the projections of the rates onto the first two components are largely distributed along the first principal component (Fig.~\ref{fig:rate}A). This and the variable amounts of sequence divergence in the regions led us to suspect the first principal component was a measure of the overall rate of sequence evolution. Plotting the first principal component against the sum of the average amino acid and indel rates as measured by substitution models revealed a strong association (Fig.~\ref{fig:rate}B). We then projected the rates along second and third principal components to determine if these higher order components contained any additional structure. The resulting distribution is roughly triangular and contains three major axes of variation, corresponding to rapid changes in the regions' proportions of acidic, glutamine, and glycine residues (Fig.~\ref{fig:rate}C-D). Inspection of regions selected along these axes confirmed the high rates of evolution of these features (Fig.~\ref{fig:rate}E-G). Furthermore, we observe a similar distribution when the rates of the order regions were projected along their second and third principal components, which suggests a lack of constraint along these axes is a general property of rapidly evolving proteins (Fig.~\ref{sfig:order}).

\subsection{A model of constrained evolution reveals patterns of conserved features}
Though the BM process permits the inference of the rates of feature evolution after accounting for the phylogenetic relationships between species, it does not directly test for their conservation. In fact, under the BM model, trait variation is unconstrained and will increase without bound over time. Instead, evidence of conservation requires comparison to a model where trait variation is constrained. A common choice for modeling the effect of selection on the evolution of a continuous trait is the Ornstein-Uhlenbeck (OU) model. The OU model is similar to the BM model where a trait accumulates variation through small, undirected steps. However, it differs in the trait is also attracted towards an optimal value where the attraction is proportional to a parameter $\alpha$ and the trait's distance from this value. Under an additional assumption of stationarity that ensures parameter identifiability and estimate consistency, the OU model is specified with three total parameters: the optimal value, the fluctuation magnitude, and the selection strength~\cite{Ho2013, Ho2014}. While the first two parameters are analogous to the root and rate parameters in the BM model, respectively, the selection strength has no equivalent.

We therefore fit OU models to the features calculated from the segments in each region to detect evidence of their conservation. A strength of probabilistic models of trait evolution is they permit the use of standard statistical methods to compare the support for different hypotheses. Thus, the goodness of fit of the OU model relative to the BM model is a measure of the evidence for selective constraint. However, because the OU model is specified with an additional parameter and has a greater capacity to fit the data, we compared their fits using the difference in the Akaike information criterion (AIC), which accounts for the model complexity, rather than the likelihood ratio. We chose not to treat these differences in a strict statistical sense but instead as qualitative ``evolutionary signatures'' as in Zarin \textit{et al.}~\cite{Zarin2019}. Accordingly, we clustered the signatures to identify patterns of feature conservation across the proteome. To ensure the clusters were enriched in regions with a high likelihood of feature conservation despite low levels of sequence identity, we restricted this analysis to regions with a minimum amount of divergence as measured by substitution models (Fig.~\ref{sfig:hierarchy_histogram}).

\begin{figure}[h!]
\includegraphics[width=\textwidth]{hierarchy/out/hierarchy.png}
\centering
\caption{\textbf{Hierarchical clustering of evolutionary signatures.}
The AIC difference between the BM and OU models is measure of their relative goodness of fit to the data with a penalty for the number of parameters in each. Larger values indicate a better fit by the OU model. Clusters are indicated by rectangles on the right. The cluster descriptions are summaries of selected significantly enriched annotations ($p < 0.01$) and do not necessarily correspond to specific terms.}
\label{fig:hierarchy}
\end{figure}

The resulting heatmap reveals patterns of similar signatures interspersed among a high background of noise (Fig.~\ref{fig:hierarchy}). However, we identify at least 12 clusters with a strong and consistent patterns of constraint, with many having significant associations with certain processes, functions, or compartments. For example, cluster B is strongly enriched in proteins associated with the Wnt signaling pathway, and cluster F contains many proteins involved in nuclear transport. (Selected annotations corresponding to the summaries in Fig.~\ref{fig:hierarchy} are given in Table~\ref{stable:annotations}.)  Though these preliminary results suggest the existence of high-level relationships between certain patterns of conserved features and function, further insight into the biophysical and mechanistic details that underpin these associations will rely on more in-depth analyses of specific clusters and the properties of their constituent IDRs.

\section{Discussion}
% I. Summary of results
%    A. In this study, we applied several phylogenetic models to IDRs to interrogate the evolution of their sequences and molecular features.
%    B. Most significantly, we demonstrate evidence of widespread evolutionary constraint in IDRs within the \textit{Drosophila} proteome
%    C. Furthermore, many IDRs can be grouped into discrete clusters on the basis of these evolutionary signatures
%    B. Most immediately, this work generates hypotheses of function related to sequence features that will permit the functional dissection of IDRs
%    C. More broadly, however, it illustrates constraint of distributed features characterizes the evolution of some IDRs across multiple systems.
%    D. However, these clusters require careful interpretation. The cluster descriptions summarize only a subset of the annotations with statistically significant enrichments. All clusters had many enriched annotations, but the summaries correspond to sets of enrichments with consisten . Because the significance tests were not corrected for multiple testing or controlled for their false discovery rate, many annotations are ,

% II. IDRs have characteristic patterns of amino acid substitution
%    A. In addition to these global patterns of conserved features, we find IDRs exhibit other characteristic patterns of evolution.
%    B. For example, a comparison of the exchangeability matrices fit to the disorder and order regions show decreased exchangeability coefficients between
%    A. Interestingly, the exchangeability matrices suggest that although there is net flux towards disorder by depleting structure-promoting residues, disordered residues are more biochemically distinct in disordered regions than in structured regions. This suggests that although disorder-promoting amino acids are often grouped into a single category and described as largely hydrophilic, they have distinct roles in disordered regions. There is evidence that highly similar residues such as glutamine and asparagine have distinct properties in disordered regions that yielded different biophysical characters of phase-separated condensates, so this aligns with other observations. Conversely, structure promoting residues are more interchangeable. One hypothesis is as these bulkier hydrophobic residues are not constrained by packing, they are free to act more generically as small hydrophobic patches that mediate interactions in SLiMs and other binding regions in IDRs. Another model for this behavior is the stickers and spacers model of phase separation.
%    A. A caveat of this analysis is its dependence on sequence alignments which were created using scoring matrices that are in turn often derived from substitution models. Some approaches attempt to minimize the impact of this circular dependency through an EM-like procedure where substitution models are fit and the sequences are re-aligned with scoring matrices derived from those models in alternating rounds until convergence. Because our analyses fit the substitution models directly from alignments, the observed patterns we observe could be artifacts of using a scoring matrix largely fit to folded domains to align IDRs
%    A. However, as the exchangeability and rate matrices broadly reproduced the trends observed in previous studies which fit amino acid substitution models to disorder regions, these observations are likely not artifactual but instead represent reproducible differences in the substitution rates between residues.
%    B. Interpretations of exchangeability and rate matrices
%        1. Disorder-enriched residues are less interchangeable and play specific roles in these regions
%        2. Disorder-depeleted residues not constrained by packing and can act more generically as small hydrophobic patches
%            i. Stickers and spacers from Holehouse
%            ii. Staller papers and other high throughput mutational scanning of ADs
%            iii. Halfmann glutamine vs asparagine paper
%        3. Note about variances of parameters as shown in CV graphs

\subsection{Disorder is correlated with many molecular features}
Though the mutations generated by evolution are not a random or exhaustive sample of sequence space, they are perturbations of a common ancestor which can reveal the relationship between a region's biophysical properties and its propensity for disorder. Thus, the analysis of feature and score contrasts is effectively a natural ``mutational scanning'' experiment. We found that disorder scores have the strongest correlations with features that measure a region's overall polarity and hydrophobicity, \textit{e.g.} fraction\_disorder and hydropathy. As the formation of a hydrophobic core is a major driving force in protein folding, this relationship is expected. However, the strength of these correlations indicate that a region's relative proportion of hydrophilic and hydrophobic residues is, to a first approximation, the largest determinant of intrinsic disorder.

Excluding these hydrophobicity-related features, the next strongest association is a negative correlation with wf\_complexity, demonstrating the predictor strongly associates low complexity with intrinsic disorder. However, sequence complexity is a statistical rather than a biophysical criterion, and while many disordered regions have low levels of complexity, some low-complexity regions, like collagen, are structured. This suggests that while disorder predictors are in general accurate classifiers of a residue's structural state, they can conflate the correlates of intrinsic disorder with their causes. Therefore, in some cases their predictions may require careful interpretation.

The remaining significant correlations are generally weak and likely reflect a partial redundancy with the more strongly correlated features discussed previously. Interestingly, though, disorder scores are weakly correlated with many motifs, with the signs largely reflecting their class. For example, the correlations with docking (DOC) and ligand binding (LIG) sites are largely negative whereas those with modification (MOD) sites are generally positive. However, this analysis does not indicate whether the predictor responds to these motifs directly or to features that are correlated with them. For example, docking and ligand binding sites are generally mediated by small hydrophobic patches, so the correlations could reflect an increase in hydrophobicity caused by an additional hydrophobic residue ``completing'' the motif rather than the motif itself. Likewise, IDRs are highly enriched in phosphorylation sites, many of which are targeted to disorder-promoting residues like serine or threonine. As the disorder scores and features are calculated at the level of regions, whose lengths can exceed 1,000 residues, this analysis is limited in its ability to distinguish these possibilities. However, a more targeted \textit{in silico} mutational analysis would yield further insights.

The GO annotation enrichment analysis indicates that proteins involved in DNA repair and extracellular structures contain a disproportionate number of regions whose disorder scores are rapidly evolving. Because the significance tests were not corrected for multiple testing or controlled for their false discovery rate, we caution against over-interpreting this result and instead consider it as a hypothesis for further investigation. In general, however, these regions with rapidly evolving disorder scores may correspond to molecular recognition features (MoRFs), which are modules in IDRs that undergo a disorder-to-order transition on binding their targets~\cite{vanderLee2014}. Because MoRFs already exist on the boundary between disorder and structure, small changes in the biophysical properties of these regions may have disproportionate effects on their structural ensembles. Thus, the most variable disorder scores may reflect instances where a mutation triggered a ``phase transition'' between largely structured or disordered native states.

\subsection{Future evolutionary analyses of IDRs will require a multimodal approach}
As discussed by Zarin \textit{et al.}, the interpretation of evolutionary signatures is complicated by several methodological limitations~\cite{Zarin2019}. For example, because IDRs are identified as contiguous segments of high predicted disorder, their boundaries are defined by adjacent structural elements. This approach can therefore split an IDR that is a single evolutionary or functional unit if it contains a semi-disordered module that scores below the threshold. Conversely, it can also merge two distinct IDRs if they are not separated by at least one folded domain. Another challenge is the significant overlap in the definitions of many features induces strong correlations that preclude straightforward quantitative manipulations or interpretations of an IDR's evolutionary signature. The original authors have since addressed this in subsequent studies by applying machine learning methods to perform feature selection or learn features directly from alignments of IDRs~\cite{Zarin2021, Lu2022}. However, integrating these methods into a unified phylogenetic comparative framework will require further effort.

By fitting the BM and OU models to molecular features calculated from alignments of IDRs, we were able to quantify the relative support for constrained and unconstrained models of IDR evolution using a statistical framework. As the BM and OU models describe the evolution of arbitrary continuous traits, a strength of this approach is its independence from assumptions about the underlying process of sequence evolution. In contrast, the previous study used simulations to generate null distributions for a model of no constraint and defined an IDR's evolutionary signature as its deviation from these distributions. However, these comparisons do not directly demonstrate evidence of stabilizing selection but instead test for differences from the null hypothesis. Thus, this approach is highly dependent on the specification and parameterization of these simulated models. Accordingly, an error in either can yield evidence of constraint for an IDR even if none of its molecular features are under selection.

However, the comparative phylogenetics methodology applied here also has limitations. As many features have strict boundaries or cannot vary continuously, they violate one or more of the underlying assumptions of the BM or OU models. Fortunately, for many features these inconsistencies likely do not seriously compromise the analysis. For example, though compositional features like fraction\_S are mathematically restricted to the interval between zero and one, they are likely constrained by much narrower selective regimes, and within these regimes, their behavior is effectively described by an OU model. For other features, however, the deviations are more consequential. For example, net\_charge and the motif features can only assume integer and non-negative values, respectively, which imposes significant restrictions on their allowed increments that are not reflected in the BM and OU models. Instead, more appropriate models for count data are birth-death processes, which are Markov chains defined on the natural numbers. However, though linear birth-death processes are well-studied and widely applied in biology~\cite{Crawford2011}, to our knowledge there are no simple parameterizations which describe a mean-reverting behavior analogous to the OU model. Thus, further theoretical developments are needed to apply birth-death processes as a model of stabilizing selection in studies of IDR evolution.

While the BM and OU models are powerful tools for studying trait evolution, their generality limits the specificity of the hypotheses they can test. In contrast, because simulation-based approaches can specify arbitrary constraints, they permit investigations of increasingly refined models of IDR evolution. We therefore view the two approaches as complementary, and propose the use of hybrid methods where test statistics are derived from phylogenetic comparative methods like the BM model, and simulations generate the null distributions for those test statistics. While simulations can eliminate specific hypotheses, the substantial resources involved both in designing them and generating samples make exhaustively testing mechanisms of feature constraint by simulation impractical. We instead recommend a workflow that begins with an analysis using general models of trait evolution to suggest specific hypotheses of constraint that are then tested with a simulation-based approach. Another hybrid method involves using sequence permutations to test for the conservation of motifs or patterned features like kappa. In this method, a comparative model like BM provides the test statistic as before, and a sample of shuffled sequences approximates the null distribution. Because the background distribution of residues is preserved, this procedure can specifically test for the conservation of motifs or patterned features in an IDR independent of the conservation of its composition.

However, as IDRs are likely subject to multiple selective pressures where the constraints on different compositional, patterned, or motif features is highly specific to each, we anticipate a range of computational and experimental methods will be needed to disentangle the complex forces driving IDR evolution. Accordingly, while these results represent a significant step forward in relating sequence to function in IDRs, further studies exploring these and other approaches will undoubtedly reveal new insights into these pervasive but poorly understood regions of proteins.

\section{Materials and methods}
\subsection{Alignment and species tree provenance}
Alignments of 8,566 single copy orthologs and the corresponding outputs of the missing data phylo-HMM were obtained from the analyses conducted in chapter~\ref{chapter:1}. Likewise, the LG consensus tree generated by the ``non-invariant, 100\% redundancy'' sampling strategy was used as the input or reference where indicated in subsequent phylogenetic analyses.

\subsection{IDR prediction and filtering}
Based on its strong performance in a recent assessment of disorder predictors, AUCPreD was chosen to identify regions with a high confidence of intrinsic disorder~\cite{Wang2016, Necci2021}. After removing the gap symbols from the sequences in the alignments, the disorder scores of each sequence were predicted individually. (Alignments 0204 and 35C2 contained sequences which exceeded the 10,000-character limit and were excluded from subsequent analyses.) The resulting scores were then aligned using the original alignment. The average score for each position was calculated using Gaussian process sequencing weighting over the LG consensus tree~\cite{Altschul1989}. Any positions inferred as ``missing'' by the missing data phylo-HMM or to the left or right of the first or last non-gap symbol, respectively, were excluded. For simplicity, the Gaussian process weights were not re-calculated from a tree pruned of the corresponding tips, and instead the weights corresponding to the remaining sequences were re-normalized. The scores at any remaining positions with gap symbols were inferred by linear interpolation from the nearest scored position.

The average disorder scores were converted into contiguous regions with the following method. Two binary masks were defined as positions where the average score exceeded high and low cutoffs of 0.6 and 0.4, respectively. The low-cutoff mask was subjected to an additional binary dilation with a structuring element of size three to merge any contiguous regions separated by a small number of positions with scores below the cutoff. ``Seed'' regions were then defined as 10 or more contiguous ``true'' positions in the high-cutoff mask, and ``disorder'' regions were obtained by expanding the seeds to the left and right until the first ``false'' position in the low-cutoff mask or the end of the alignment. ``Order'' regions were taken as the complement of the disorder regions in each alignment.

The regions were filtered with the following criteria. First, segments with non-standard amino acid symbols, which overlapped with any position labeled as ``missing'' by the missing data phylo-HMM, or whose number of non-gap symbols was below a length cutoff of 30 residues were removed. Regions whose remaining segments failed the set of phylogenetic diversity criteria detailed in Table~\ref{stable:diversity_criteria} were excluded. The final set contained 11,445 and 14,927 disorder and order regions, respectively, from 8,466 distinct alignments.

\subsection{Fitting substitution models and trees}
To fit amino acid substitution matrices to disorder and order regions, 25 meta-alignments were constructed by randomly sampling 100,000 columns from the respective subsets in the filtered regions. To determine the effect of gaps, the maximum fraction of gaps was set at 0, 50, 100\%. The combination of the region types and sampling strategy yielded six different sets of meta-alignments. A GTR20 substitution model with four FreeRate categories and optimized state frequencies was fit to each meta-alignment using IQ-TREE 1.6.12~\cite{Nguyen2014}. Exchangeability and rate coefficients were normalized, so the average rate of each model was equal to 1. Because exchangeability and rate coefficients are highly correlated across meta-alignments of the same region type, all figures are derived from the maximum 50\% gap fraction meta-alignment sets unless otherwise noted (Fig.~\ref{sfig:heatmap_ematrix}-\ref{sfig:heatmap_corr}).

To obtain estimates of the average substitution rates in each region, separate amino acid and indel models were fit to each alignment. For the amino acid substitution models, the columns in the alignments were manually segregated into disorder and order partitions using the regions derived from the AUCPreD scores. However, to prevent poor fits from a lack of data, a partition was created only if it contained a minimum of 20 sequences with at least 30 non-gap symbols. If one partition met these conditions but the other did not, the disallowed partition was consolidated into the allowed one. If neither partition passed, the alignment was skipped. These rules ensured that the regions represented in the filtered set were fit with substitution models which were concordant with their predicted disorder states. Trees were fit to each partition using an invariant and four discrete gamma rate categories using IQ-TREE 1.6.12~\cite{Yang1994}. The disorder partition used a substitution model derived from the average of the state frequencies and rate coefficients fit to the 50\% gap fraction meta-alignment sets sampled from the disorder regions. The order partition used the LG substitution model~\cite{Le2008}. To prevent overfitting of branch lengths, the trees were restricted to scaled versions of the reference species tree using the -{}-blscale option.

As inference with models that allow insertions and deletions of arbitrary lengths is computationally intractable, a more heuristic approach was taken to quantify the amount of evolutionary divergence resulting from indels in the alignments. For a given alignment, all contiguous subsequences of gap symbols with unique start or stop positions in any sequence were defined as binary characters. Then for each character, a sequence was coded with the symbol 1 if it contained that character or that character was nested in another contiguous subsequence of gap symbols in the sequence and the symbol 0 otherwise. GTR2 models with optimized state frequencies and ascertainment bias correction were fit to the resulting character alignments. A discrete gamma rate category was added for every five character columns, up to a maximum of four. To prevent overfitting of branch lengths, the trees were restricted to scaled versions of the reference species tree using the -{}-blscale option.

Because the rate and branch lengths of a phylogenetic substitution model always appear as products in the likelihood expression, they are not jointly identifiable parameters. Instead, the rate is conventionally taken as equal to one (with inverse count units), and the branch lengths are expressed in terms of the expected number of substitution events per column. For models with multiple rate categories, the equivalent condition is that the mean of the prior distribution over the rate categories is equal to one. This effectively makes each rate category a scaling factor of the branch lengths. The inferred rate of a column, calculated as the mean of the posterior distribution over the rate categories, is therefore relative to the average across all columns in the alignment. An absolute measure of the evolutionary divergence of a column can be obtained by multiplying the inferred rate by the total branch length of the tree. However, as the alignments contain variable numbers of species, this total branch length represents the contribution of both the rate and the tree topology. To normalize for this effect, the total branch length for each tree fit to an alignment was divided by the total branch length of the reference species tree including only the species in that alignment. The reported substitution rate is therefore a product of this scaling factor and the inferred column rate. The average amino acid or indel substitution rate for a region was calculated as the mean of the respective rate across all columns. Because the indel rates were associated with columns in character alignments, they were mapped back to the original sequence alignment by assigning half of a character's rate to its start and stop positions. Since indel models with limited data were prone to overfitting, rates obtained from character alignments with fewer than five columns were set to zero.

\subsection{Definition and calculation of features}
Features were calculated as in Zarin \textit{et al.} with the following modifications~\cite{Zarin2019}. The regular expression for polar residue fraction was [QNSTCH], which, in contrast to the original study, excludes glycine residues. Additionally, length, expressed in log scale, was replaced with a feature proportional radius of gyration for an excluded-volume polymer~\cite{Flory1949}. Because the radii of gyration of chemically denatured proteins closely match the values expected for equivalent random coils, we felt this feature would better capture the relationship between an IDR's length and its biophysical properties~\cite{Kohn2004}. Finally, several motifs from ELM were replaced with their metazoan counterparts or updated versions of the same entries~\cite{Kumar2021}. These differences are noted in the supplementary data. Furthermore, unlike the previous work, motifs were left as counts and not normalized to the proteome-wide average. Kappa, omega, SCD, hydropathy, PPII propensity, and Wootton-Federhen sequence complexity were calculated with localCIDER 0.1.19~\cite{Holehouse2017}. Isoelectric point was calculated with the Python package isoelectric, which is available on PyPI or at \url{https://isoelectric.org/}~\cite{Kozlowski2016}. Otherwise, features were implemented with custom code. A full list of features and their definitions is given in Table~\ref{stable:features} and Table~\ref{stable:regexes}.

\subsection{Brownian motion and Ornstein-Uhlenbeck analyses}
Brownian motion (BM) model parameters were calculated with two methods. The first used Felsenstein's contrasts algorithm to efficiently calculate roots and contrasts for the disorder scores and features of each region~\cite{Felsenstein1973, Felsenstein1981}. Rates were calculated as the mean of the squares of the contrasts. Though these values are unbiased, they are not maximum likelihood estimates and are inappropriate for use with the Akaike information criterion (AIC)~\cite{Akaike1974}. Thus, they were used for analyses involving only the BM model. For comparison with the Ornstein-Uhlenbeck (OU) model, the BM parameters were calculated by maximizing the likelihood. The OU model parameters were also calculated via maximum likelihood estimation. To ensure parameter identfiability and estimate consistency, the root was treated as a random variable~\cite{Ho2013, Ho2014}. Thus, the covariance matrix, $V$, was parameterized as $V_{ij} = e^{\alpha d_{ij}}$ where $d_{ij}$ is the tree distance between tips $i$ and $j$, and $\alpha$ is the selection strength~\cite{Ho2013}.

The AICs were calculated for the models of each feature using their maximized likelihoods and two and three parameters for the BM and OU models, respectively. The pairwise differences in the AICs yielded a vector with 82 components, each representing the relative goodness of fit of the OU model over the BM model after accounting for their difference in complexity. The vectors were clustered using the correlation distance metric and the UPGMA algorithm. Clusters were manually chosen for subsequent GO analyses. To enrich these clusters for regions with a high likelihood of feature conservation despite low levels of sequence identity, this analysis was restricted to 7,645 regions whose amino acid and indel substitution rates exceeded 1 and 0.1, respectively (Fig.~\ref{sfig:hierarchy_histogram}).

\subsection{GO term analyses}
The 2022-03-22 go-basic release of the Gene Ontology was obtained from the GO Consortium website~\cite{Ashburner2000, 2020GOConsortium}. The gene association file for the 2022\_02 release of the \textit{D. melanogaster} genome annotation was obtained from FlyBase~\cite{Gramates2022}. Obsolete annotations were dropped, and the remaining annotations were filtered by qualifiers and evidence code. The allowed qualifiers were ``enables,'' ``contributes\_to,'' ``involved\_in,'' ``located\_in,'' ``part\_of,'' and ``is\_active\_in.'' The allowed evidence codes were all experimental sources, traceable author statement (TAS), and inferred by curator (IC). The annotations were propagated up the ontology graph and joined with the region sets, so every annotation associated with a gene was associated with the regions derived from that gene. \textit{P}-values were calculated with exact hypergeometric probabilities with regions considered as the sampling unit. For the disorder score analysis, the reference set was the filtered regions, and the enrichment set was regions in the upper decile of the score rate distribution (Fig.~\ref{sfig:scores_histogram}). For the cluster analysis, the reference set was the filtered regions after the additional filtering by substitution rates, and the enrichments sets were the regions in each cluster.

\subsection{Code and data availability}
\begin{sloppypar}
The code used to produce the results and analyses is available at \url{https://github.com/marcsingleton/IDR_evolution2023}. The following Python libraries were used: matplotlib~\cite{Hunter2007}, NumPy~\cite{Harris2020}, pandas~\cite{McKinney2010}, SciPy~\cite{Virtanen2020}, and scikit-learn~\cite{Pedregosa2011}. Relevant output files are available in the supporting information. There is no primary data associated with this manuscript. All primary data are available from publicly accessible sources described in their corresponding sections.
\end{sloppypar}
