\begin{abstract}
\noindent
Here's where my abstract will go!
\end{abstract}

\section{Introduction}
Here's where my introduction will go!

\section{Results}
\begin{figure}[h!]
\includegraphics[width=\textwidth]{substitution/out/substitution.png}
\centering
\caption{\textbf{Amino acid substitution models fit to disorder and order regions.}
\textbf{(A)} Amino acid frequencies of substitution models. Amino acid symbols are ordered by their enrichment in disorder regions, calculated as the disorder-to-order ratio of their frequencies. Error bars represent standard deviations over models fit to different meta-alignments ($n = 25$). \textbf{(B-C)} Exchangeability coefficients of disorder and order regions, respectively. \textbf{(D)} $\log_{10}$ disorder-to-order ratios of exchangeability coefficients. \textbf{(E-F)} Rate coefficients of disorder and order regions, respectively. The vertical and horizontal axes indicate the initial and target amino acids, respectively. \textbf{(G)} $\log_{10}$ disorder-to-order ratios of rate coefficients.}
\label{fig:substitution}
\end{figure}

\section{Discussion}
Here's where my discussion will go!

\section{Materials and methods}

\subsection{Alignment and tree generation}
Alignments of single copy orthologs and the outputs of the missing data phylo-HMM were obtained from the analyses conducted in chapter~\ref{chapter:1}. Likewise, the LG consensus tree generated by the ``non-invariant, 100\% redundancy'' sampling strategy was used as the input or reference where indicated in subsequent phylogenetic analyses.

\subsection{IDR prediction and filtering}
AUCPreD was used to predict the disorder score of each residue of each sequence in the alignments after the gap symbols were removed~\cite{Wang2016}. The resulting scores were then aligned using the original alignment. The average score for each position was calculated using Gaussian process sequencing weighting over the LG consensus tree~\cite{Altschul1989}. Any positions inferred as ``missing'' by the missing data phylo-HMM or to the left or right of the first or last non-gap symbol, respectively, were excluded. For simplicity, the Gaussian process weights were not re-calculated from a tree pruned of the corresponding tips, and instead the weights corresponding to the remaining sequences were re-normalized. The scores at any remaining positions with gap symbols were inferred by linear interpolation from the nearest scored position.

The average disorder scores were converted into contiguous regions with the following method. Two binary masks were defined as positions where the average score exceeded high and low cutoffs of 0.6 and 0.4, respectively. The low-cutoff mask was subjected to an additional binary dilation with a structuring element of size three to merge any contiguous regions separated by a small number of positions with scores below the cutoff. ``Seed'' regions were then defined as 10 or more contiguous ``true'' positions in the high-cutoff mask, and ``disorder'' regions were obtained by expanding the seeds to the left and right until the first ``false'' position in the low-cutoff mask or the end of the alignment. ``Order'' regions were taken as the complement of the disorder regions in each alignment.

The regions were filtered with the following criteria. First, segments with non-standard amino acid symbols, which overlapped with any position labeled as ``missing'' by the missing data phylo-HMM, or whose number of non-gap symbols was below a length cutoff were removed. Regions whose remaining segments satisfied the set of phylogenetic diversity criteria detailed in Table~\ref{stable:diversity_criteria} were included in the final set. The length cutoffs were set to 30, 60, and 90 residues, which generated three distinct sets of regions.

\subsection{Fitting amino substitution models}
To fit amino acid substitution matrices to disorder and order regions, 25 meta-alignments were constructed by randomly sampling 100,000 columns from the respective subsets in the filtered regions derived from the 30 residue length cutoff. To determine the effect of gaps, the maximum fraction of gaps was set at 0, 50, 100\%. The combination of the region types and sampling strategy yielded six different sets of meta-alignments. A GTR20 substitution model with four FreeRate categories and optimized state frequencies was fit to each meta-alignment using IQ-TREE 1.6.12~\cite{Nguyen2014}. Exchangeability and rate coefficients were normalized to make the average rate of each model equal to 1. All figures are derived from the maximum 50\% gap fraction meta-alignment set unless otherwise noted.

\subsection{Definition and calculation of features}
Features were calculated as in Zarin \textit{et al.} with the following modifications~\cite{Zarin2019}. The regular expression for polar residue fraction was [QNSTCH], which, in contrast to the original study, excludes glycine residues. Additionally, length, expressed in log scale, was replaced with a feature proportional radius of gyration for an excluded-volume polymer~\cite{Flory1949}. Because the radii of gyration of chemically denatured proteins closely match the values expected for equivalent random coils, we felt this feature would better capture the relationship between an IDR's length and its biophysical properties~\cite{Kohn2004}. Finally, several motifs from ELM were replaced with their metazoan counterparts or updated versions of the same entries~\cite{Kumar2021}. These differences are noted in the supplementary data. Kappa, omega, SCD, hydropathy, PPII propensity, and Wootton-Federhen sequence complexity were calculated with localCIDER 0.1.19~\cite{Holehouse2017}. Isoelectric point was calculated with the Python package isoelectric, which is available on PyPI or at \url{https://isoelectric.org/}~\cite{Kozlowski2016}. Otherwise, features were calculated with custom code. A full list of features and their definitions is given in Table~\ref{stable:features} and Table~\ref{stable:regexes}.

\subsection{Brownian motion and Ornstein-Uhlenbeck analyses}

\subsection{Simulation analyses}

\subsection{GO term analyses}

\subsection{Code and data availability}
The code used to produce the results and analyses is available at \url{https://github.com/marcsingleton/IDR_evolution2023}. The following Python libraries were used: matplotlib, NumPy, pandas, SciPy, and scikit-learn~\cite{Hunter2007, Harris2020, McKinney2010, Virtanen2020, Pedregosa2011}. Relevant output files are available in the supporting information. There is no primary data associated with this manuscript. All primary data are available from publicly accessible sources described in their corresponding sections.