\begin{abstract}
\noindent
Here's where my abstract will go!
\end{abstract}

\section{Introduction}
Here's where my introduction will go!
% I. IDRs, proteins without defined structures, are prevalent in eukaryotic proteins and play key roles in many processes
% II. Sequence-function relationships are unclear in IDRs
% III. Proteome-wide patterns of conserved features linked to function
%    A. Past studies have demonstrated evidence of constraint in specific proteins for specific properties
%    B. Additionally other studies have shown importance of short linear motifs in IDRs
%        1. Can be identified as short segments of conservation in otherwise highly diverged alignments of IDRs
%        2. However, these segments comprise a small fraction of disordered regions, so
%    C. Recent study first to demonstrate patterns of selection called evolutionary signatures in IDRs across entire proteome using a set of 82 molecular features
%    D. Furthermore, by clustering IDRs on by these evolutionary signatures, they identified groups of IDRs evolving under similar sets of constraints. Furthermore they identified functional associations in these clusters which can be used to generate hypotheses of function for previously uncharacterized proteins
% IV. Unclear if these patterns are specific to yeast or representative of IDR evolution across the tree of life
%    A. Expand analyses to another model organism
%        1. Drosophila is key model for genetic and development research with abundant genomic resources across many lineages in the radiation
%        2. Furthermore as multicellular organism with complex developmental pathways, the results would be particularly relevant for understanding the role of IDRs in complex signaling pathways
%        3. Also relevant for human health!
%    B. Results would have significant impact on interpretation
%        1. Conserved distributed properties are a yeast phenomena
%        2. Patterns of conservation are yeast phenomena
%        3. Global taxonomy of disordered regions similar to Pfam
% IV. Explicit phylogenetic models of trait evolution can reveal patterns of evolution with fewer assumptions
%    A. Previous analyses compared observed distributions of features in observed IDRs to the distributions generated by simulated counterparts
%    B. Differences were interpreted as evidence of selection on features
%        1. However, evidence of deviation from the null model rather than evidence of constraint necessarily
%        2. Null distribution generated by a model of evolution where the composition of sequences stochastically evolve towards a specific equilibrium distribution
%        3. Previous studies have identified IDRs with varying compositional preferences, so a single equilibrium distribution is unlikely to globally describe observed patterns of sequence evolution in IDRs
%    C. Thus in this work we explore use of various statistical approaches that directly measure rates of change and constraint of molecular features indepedent of specific model of sequence evolution
%        1. Place these evolutionary analyses in a statistical phylogenetic framework
%        2. Though these models have own assumptions, they are macroscopic descriptions of continuous trait evolution and therefore independent of any assumptions about evolution at the level sequences
%        3. Still have their own assumptions of course, but more general and robust to deviations
%        4. We view these approaches as complementary, each having different trade-offs between specificity and generality of testing specific hypotheses of disorder evolution
%        5. We also explore hybrid approaches where the evolutionary rates obtained from phylogenetic models are used as a statistic
% V. By applying these approaches to set of IDRs identified in Drosophila genomes, we observe...

\section{Results}
\subsection{IDRs are shorter and more divergent than non-disordered regions}
We identified regions with high levels of inferred intrinsic disorder in over 8,500 alignments of single copy orthologs from 33 species in the \textit{Drosophila} genus using the disorder predictor AUCPreD~\cite{Wang2016}. To highlight the unique features of IDR evolution in subsequent analyses, we also extracted a complementary set of regions with low levels of inferred disorder. Both sets were filtered on several criteria, including the lengths of their sequences and their phylogenetic diversity, which yielded 11,445 and 14,927 regions, respectively, from 8,466 unique alignments. In the subsequent discussion, we refer to these sets as the ``disorder'' and ``order'' regions, respectively. To investigate the broad differences in basic sequence statistics between the two region sets, we first generated histograms from the average length of each region (Fig.~\ref{sfig:regions}). Although both distributions span several orders of magnitude, the order regions are generally longer than the disorder regions, with means of 105 and 245 residues, respectively. We then quantified the sequence divergence in each region by fitting phylogenetic trees to the alignments using amino acid and indel substitution models. The average rates of substitution are significantly larger in the disorder regions, demonstrating that while both sets contain conserved and divergent regions, IDRs are enriched in more rapidly evolving sequences (Fig.~\ref{sfig:regions}).

\subsection{Disorder regions have distinct patterns of residue substitution}
To gain insight into the substitution patterns of amino acid residues in disorder and order regions, we fit substitution models to meta-alignments sampled from the respective sets of filtered regions. These models are probabilistic descriptions of sequence evolution and are parameterized in terms of the one-way rates of change from one residue to another. Thus, the rates are not necessarily equal for a given pair of residues when the initial and target residues are swapped. For example, the rate of change of valine to tryptophan can be distinct from that of tryptophan to valine. In practice, substitution models are typically constrained to fulfill a condition called time-reversibility, as this converts a difficult multivariate optimization of the tree's branch lengths into a series of simpler univariate optimizations~\cite{Felsenstein1981}. A common method for fulfilling this condition is parameterizing the model in terms of a frequency vector, $\boldsymbol \pi$, and an exchangeability matrix, $S$. The frequency vector determines the model's expected residue frequencies at equilibrium, meaning the model dictates that all sequences eventually approach this distribution, no matter their initial composition. The exchangeability matrix is symmetric ($s_{ij} = s_{ji}$) and encodes the propensity for two residues to interconvert. Because the rate of change from residue $i$ to residue $j$ is given by $r_{ij} = s_{ij}\pi_{j}$, higher exchangeability coefficients yield higher rates of conversion. Thus, exchangeability coefficients are frequently interpreted as a measure of biochemical similarity between residues.

\begin{figure}[h!]
\includegraphics[width=\textwidth]{substitution/out/substitution.png}
\centering
\caption{\textbf{Amino acid substitution models fit to disorder and order regions.}
\textbf{(A)} Amino acid frequencies of substitution models. Amino acid symbols are ordered by their enrichment in disorder regions, calculated as the disorder-to-order ratio of their frequencies. Error bars represent standard deviations over models fit to different meta-alignments ($n = 25$). \textbf{(B-C)} Exchangeability coefficients of disorder and order regions, respectively, averaged over meta-alignments. \textbf{(D)} $\log_{10}$ disorder-to-order ratios of exchangeability coefficients. \textbf{(E-F)} Rate coefficients of disorder and order regions, respectively, averaged over meta-alignments. The vertical and horizontal axes indicate the initial and target amino acids, respectively. \textbf{(G)} $\log_{10}$ disorder-to-order ratios of rate coefficients.}
\label{fig:substitution}
\end{figure}

To highlight the differences in patterns of residue substitution between the disorder and order regions, the parameters in each model are directly compared in Fig.~\ref{fig:substitution}, beginning with the frequency vectors. The disorder regions show an enrichment of ``disorder-promoting'' residues such as serine, proline, and alanine, and a depletion of hydrophobic and bulky residues such as trytophan and phenylalanine (Fig.~\ref{fig:substitution}A). The exchangeability matrices fit to the disorder and order regions have similar overall patterns of high and low coefficients (Fig.~\ref{fig:substitution}B-C). However, the log ratios of the disorder to the order exchangeability coefficients show clear differences within and between the disorder-enriched and -depleted residues. In disorder regions, disorder-enriched residues are less exchangeable with each other, whereas disorder-depleted residues are more exchangeable with each other and with disorder-enriched residues (Fig.~\ref{fig:substitution}D). Likewise, we observe a trend in the log ratios of the rate coefficients where the coefficients above the diagonal are generally positive and those below the diagonal are generally negative. As the coefficients model the one-way rate of substitutions between residues with the vertical and horizontal axes indicating the initial and target residues, respectively, this suggests a net flux towards a more disorder-like composition.

% RESULTS
% III. Brownian motion score analyses
%     A. Though the substitution models reveal specific patterns of evolution at the level of individual residues in IDRs, the large amounts of sequence divergence in many orthologous IDRs implies their evolution is not well-described by fine-scale models of residue substitution. Given the growing evidence that IDRs are constrained to conserve distributed properties, we instead turned towards characterizing their evolution as whole. Thus, for each sequence segment composing a region we calculated 82 disorder-associated ``molecular features'' obtained from the analyses of IDRs found in the yeast proteome.
%     B. However, before conducting an in-depth analysis of these features, we examined the disorder score traces in greater detail and were struck by the significant variability between species.
%         1. For each residue in the input sequence, AUCPreD returns a score between 0 and 1 where higher values indicate higher confidence in a prediction of intrinsic disorder. In some alignments, the scores would vary by nearly this entire range at a given alignment position, even when there was a relatively high level of sequence identity (Fig.~\ref{fig:scores}A).
%     A. To better understand the relationship of this variability to changes in the regions' biophysical properties, we sought to correlate the average disorder score of the segments in a region with their molecular features. However, because the sequences are the product of evolution and related by a phylogenetic tree, they are not independent, making them unsuitable for direct use in many standard statistical procedures. We instead applied the method of contrasts, which takes differences between adjacent nodes in the tree to generate a set contrasts for both the scores and the features. Under some general assumptions of the underlying evolutionary process, contrasts are independent and identically-distributed and are therefore appropriate for use in correlation analyses. The resulting feature contrasts have varying degrees of correlation with the score contrasts (Fig.~\ref{fig:scores}B-C). Some, like isopoint are uncorrelated, but most are significantly, if weakly, correlated. In general, the strongest correlations are observed for features which have a direct biophysical relationship to the presence of absence of disorder, such as hydrophobicity or disorder_fraction. Interestingly, the correlations with many motifs were statistically significant, though small in magnitude relative to the non-motif features. However, a more detailed analysis of this observation is presented in the discussion.
%     D. Though this analysis is limited by the resolution of the regions, whose lengths can exceed 1,000 residues, it indicates the prediction algorithm is sensitive to these features in bulk.
%          1. Some more observations of the overall trends ...
%     E. Interestingly, the correlations with many motifs were statistically significant, though small in magnitude relative to the non-motif features.
%          1. Look into motif regexes and see if ones with highest correlations have any common features
%     F. Next, to determine if regions with rapidly evolving disorder scores were enriched in any processes, we extracted the upper decile of their rate distribution
%          1. Many chromatin and DNA related features

% IV. Brownian motion feature analyses
%     A. Having calculated features for all IDRs in our data set, we were curious the distribution contained any global structure which would allow the identification of classes with distinct biophysical or functional properties
%     A. Feature roots
%         1. Discuss differences in scaling and need for z-scores
%         2. Discussion of sample regions in figure? (need to add)
%     B. The phylogenetic contrasts algorithm also permits an estimate of the rate of feature evolution, which is effectively the variance of the distribution of features across all sequences after correcting for the phylogenetic relationships between them
%         1. PC1 correlates with rate
%         2. PC2 and PC3 show three major axes of unconstrained variation: Q-rich, G-rich, acidic

% V. BM-OU analyses
%    A. While these analyses measure the rates of change of features after accounting the the phylogenetic relationships between sequences, they do not directly test for conservation of features
%         1. In fact, under the BM model, trait variation is unconstrained and will increase without bound over time
%         2. Instead, demonstrating evidence of conservation requires a comparison to a model where feature evolution is constrained.
%         3. A common choice is the Ornstein-Uhlenbeck (OU) model
%             i. Some explanation about how it works

% CONCLUSION
% I. Summary of results
%    A. In this study we showed...
%    B. IDRs have patterns of evolution distinct from folded domains
%    C. Elevated rates of residue substitution and indels
% II. General caveats

% II. Exchangeability and rate matrices
%    A. Reproduced trends seen in previous substitution models fit to disorder regions
%    B. Interpretations of exchangeability and rate matrices
%        1. Disorder-enriched residues are less interchangeable and play specific roles in these regions
%        2. Disorder-depeleted residues not constrained by packing and can act more generically as small hydrophobic patches
%            i. Stickers and spacers from Holehouse
%            ii. Staller papers and other high throughput mutational scanning of ADs
%            iii. Halfmann glutamine vs asparagine paper
%        3. Note about variances of parameters as shown in CV graphs

% III. Disorder score analyses
%    A. Comparison of features and disorder scores acts as a kind of ``mutational scanning'' experiment [citation to Eddy?]
%        1. Contrasts of features and disorder scores shows the disorder predictor is sensitive to most features used in this analyses
%        2. However, highest correlations to expected features like fraction_disorder and hydropathy
%        3. Interestingly, negatively correlated with wf_complexity. Many disordered regions have low levels of sequence complexity. However this is a statistical rather than a biophysical criterion, and some low complexity regions are structured, like collagen. This indicates that these models can confuse correlation and causation during training, and their representations should be interpreted carefully.
%        1. Furthermore, resolution of analyses limited because features calculated at level of regions rather than a fixed window around each residue.
%    B. GO annotation enrichment analysis suggests disorder scores undergo rapid evolution in proteins involved in DNA repair and found in extracellular structure
%        1. Not corrected for multiple testing nor is the false discovery rate controlled, so should be considered as hypotheses for further investigation
%        2. However, this result is at least consistent with other known functions of disorder
%            i. Interactions with extracellular components and signaling, both highly enriched in disorder?
%            ii. These regions could contain MoRFs, so they are already on the boundary between folded and disorder
%            iii. As evolution optimizes these interactions, small changes in the biophysical properties of these regions may have disproportionate effects on their structural ensembles
%            iv. The variable scoring by the disorder predictor may reflect this kind of ``phrase transition''

% IV. Feature analyses
%    A. Though no discrete clusters are visible, only captures x% of the overall variation, so significant amounts of variation not represented in these axes
%    B. Hierarchical clustering of the rates reveals additional structure.
%    C. Other non-linear dimensionality reductions like UMAP be more successful at preserving the fine details of the high-dimensional distributions in the low-dimensional space
%        1. However, these methods designed primarily for visualization and as result can significant distort the global structure of high-dimensional data to preverse its local topology. Thus, these embeddings are generally not suitable for downstream analyses.

% V. OU analyses
%    A. Discussion of issues with statistical inference and need for simulations to address inference issues with OU model
%        1. Lack of power or prone to type I errors?
%    B. Better models for handling motif counts
%        1. Unlike other features where small changes in a sequence general yield small changes in their value, the complexity of the regular expressions made them highly sensitive to small perturbations.
%    C. Need for comparisons with simulations with amino acid substitution and indel models
%        1. Weaknesses are explicit assumptions about substitution and indel processes and high computational cost
%        2. Is also a strength because allows testing of increasingly specific hypotheses for IDR evolution
%        3. Complementary approach -- fewer assumptions about description of evolution but cannot be tailored to specific hypotheses
%        4. Hybrid approach using comparisons of rates as the statistical test and simulations to generate the null distribution
%    D. Mention other distribution free approaches

% VI. Big picture and wrap up
%     A. Experimental swaps of IDRs in similar classes
%     B. What are features?
%         - Contrastive learning (Moses)
%         - Prediction on gold standard data sets
%         - Embeddings from LLMs trained on large data sets

\begin{figure}[h!]
\includegraphics[width=\textwidth]{scores/out/scores_07E3.png}
\centering
\caption{\textbf{Analyses of disorder scores.}
\textbf{(A)} Example region in the alignment of the sequences in orthologous group 07E3 with their corresponding disorder scores. Higher scores indicate a higher probability of disorder. Disorder score traces are colored by the position of its associated species on the phylogenetic tree. \textbf{(B)} Correlations between disorder scores and feature contrasts in regions. Asterisks indicate statistically significant correlations as computed by permutation tests ($p < 0.001$). \textbf{(C-D)} Example scatter plots showing correlations plotted in panel B. \textbf{(E)} GO term analysis of regions with rapidly evolving disorder scores. Only terms where $p < 0.001$ are shown.}
\label{fig:scores}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=\textwidth]{brownian/out/root.png}
\centering
\caption{\textbf{PCA of disorder regions' feature roots.}
\textbf{(A)} The first two PCs of the disorder regions' feature root distributions. The explained variance percentage of each component is indicated in parentheses in the axis labels. \textbf{(B)} The same plot as panel A with the projections of original features onto the components shown as arrows. Only the 16 features with the largest projections are shown. Scaling of the arrows is arbitrary.}
\label{fig:root}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=\textwidth]{brownian/out/rate.png}
\centering
\caption{\textbf{PCA of disorder regions' feature rates.}
\textbf{(A)} The first two PCs of the disorder regions' feature rate distributions. The explained variance percentage of each component is indicated in parentheses. \textbf{(B)} Scatter plot of the disorder regions' feature rates along the first PC against the sum of the average amino acid and indel substitution rates in those regions. \textbf{(C)} The second and third PCs of the disorder regions' feature rate distributions. The explained variance percentage of each component is indicated in parentheses in the axis labels. \textbf{(D)} The same plot as panel C with the projections of original rates onto the components shown as arrows. Only the 16 features with the largest projections are shown. Scaling of the arrows is arbitrary. \textbf{(E-F)} Example alignments of disorder regions from the orthologous groups 0A8A, 3139, 04B0, respectively. The colored bars on the left indicate the hexbin containing that region in panel C.}
\label{fig:rate}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=\textwidth]{hierarchy/out/hierarchy.png}
\centering
\caption{\textbf{Hierarchical clustering of evolutionary signatures.}
The AIC difference between the BM and OU models is measure of their relative goodness of fit to the data with a penalty for the number of parameters in each. Larger values indicate a better fit by the OU model.}
\label{fig:hierarchy}
\end{figure}

\section{Discussion}
Here's where my discussion will go!

\section{Materials and methods}

\subsection{Alignment and species tree provenance}
Alignments of 8,566 single copy orthologs and the corresponding outputs of the missing data phylo-HMM were obtained from the analyses conducted in chapter~\ref{chapter:1}. Likewise, the LG consensus tree generated by the ``non-invariant, 100\% redundancy'' sampling strategy was used as the input or reference where indicated in subsequent phylogenetic analyses.

\subsection{IDR prediction and filtering}
Based on its strong performance in a recent assessment of disorder predictors, AUCPreD was chosen to identify regions with a high confidence of intrinsic disorder~\cite{Wang2016, Necci2021}. After removing the gap symbols from the sequences in the alignments, the disorder scores of each sequence were predicted individually. (Alignments 0204 and 35C2 contained sequences which exceeded the 10,000-character limit and were excluded from subsequent analyses.) The resulting scores were then aligned using the original alignment. The average score for each position was calculated using Gaussian process sequencing weighting over the LG consensus tree~\cite{Altschul1989}. Any positions inferred as ``missing'' by the missing data phylo-HMM or to the left or right of the first or last non-gap symbol, respectively, were excluded. For simplicity, the Gaussian process weights were not re-calculated from a tree pruned of the corresponding tips, and instead the weights corresponding to the remaining sequences were re-normalized. The scores at any remaining positions with gap symbols were inferred by linear interpolation from the nearest scored position.

The average disorder scores were converted into contiguous regions with the following method. Two binary masks were defined as positions where the average score exceeded high and low cutoffs of 0.6 and 0.4, respectively. The low-cutoff mask was subjected to an additional binary dilation with a structuring element of size three to merge any contiguous regions separated by a small number of positions with scores below the cutoff. ``Seed'' regions were then defined as 10 or more contiguous ``true'' positions in the high-cutoff mask, and ``disorder'' regions were obtained by expanding the seeds to the left and right until the first ``false'' position in the low-cutoff mask or the end of the alignment. ``Order'' regions were taken as the complement of the disorder regions in each alignment.

The regions were filtered with the following criteria. First, segments with non-standard amino acid symbols, which overlapped with any position labeled as ``missing'' by the missing data phylo-HMM, or whose number of non-gap symbols was below a length cutoff of 30 residues were removed. Regions whose remaining segments failed the set of phylogenetic diversity criteria detailed in Table~\ref{stable:diversity_criteria} were excluded. The final set contained 11,445 and 14,927 disorder and order regions, respectively, from 8,466 distinct alignments.

\subsection{Fitting substitution models and trees}
To fit amino acid substitution matrices to disorder and order regions, 25 meta-alignments were constructed by randomly sampling 100,000 columns from the respective subsets in the filtered regions. To determine the effect of gaps, the maximum fraction of gaps was set at 0, 50, 100\%. The combination of the region types and sampling strategy yielded six different sets of meta-alignments. A GTR20 substitution model with four FreeRate categories and optimized state frequencies was fit to each meta-alignment using IQ-TREE 1.6.12~\cite{Nguyen2014}. Exchangeability and rate coefficients were normalized, so the average rate of each model was equal to 1. Because exchangeability and rate coefficients are highly correlated across meta-alignments of the same region type, all figures are derived from the maximum 50\% gap fraction meta-alignment sets unless otherwise noted (Fig.~\ref{sfig:heatmap_ematrix}-\ref{sfig:heatmap_corr}).

To obtain estimates of the average substitution rates in each region, separate amino acid and indel models were fit to each alignment. For the amino acid substitution models, the columns in the alignments were manually segregated into disorder and order partitions using the regions derived from the AUCPreD scores. However, to prevent poor fits from a lack of data, a partition was created only if it contained a minimum of 20 sequences with at least 30 non-gap symbols. If one partition met these conditions but the other did not, the disallowed partition was consolidated into the allowed one. If neither partition passed, the alignment was skipped. These rules ensured that the regions represented in the filtered set were fit with substitution models which were concordant with their predicted disorder states. Trees were fit to each partition using an invariant and four discrete gamma rate categories using IQ-TREE 1.6.12~\cite{Yang1994}. The disorder partition used a substitution model derived from the average of the state frequencies and rate coefficients fit to the 50\% gap fraction meta-alignment sets sampled from the disorder regions. The order partition used the LG substitution model~\cite{Le2008}. To prevent overfitting of branch lengths, the trees were restricted to scaled versions of the reference species tree using the -{}-blscale option.

As inference with models that allow insertions and deletions of arbitrary lengths is computationally intractable, a more heuristic approach was taken to quantify the amount of evolutionary divergence resulting from indels in the alignments. For a given alignment, all contiguous subsequences of gap symbols with unique start or stop positions in any sequence were defined as binary characters. Then for each character, a sequence was coded with the symbol 1 if it contained that character or that character was nested in another contiguous subsequence of gap symbols in the sequence and the symbol 0 otherwise. GTR2 models with optimized state frequencies and ascertainment bias correction were fit to the resulting character alignments. A discrete gamma rate category was added for every five character columns, up to a maximum of four. To prevent overfitting of branch lengths, the trees were restricted to scaled versions of the reference species tree using the -{}-blscale option.

Because the rate and branch lengths of a phylogenetic substitution model always appear as products in the likelihood expression, they are not jointly identifiable parameters. Instead, the rate is conventionally taken as equal to one (with inverse count units), and the branch lengths are expressed in terms of the expected number of substitution events per column. For models with multiple rate categories, the equivalent condition is that the mean of the prior distribution over the rate categories is equal to one. This effectively makes each rate category a scaling factor of the branch lengths. The inferred rate of a column, calculated as the mean of the posterior distribution over the rate categories, is therefore relative to the average across all columns in the alignment. An absolute measure of the evolutionary divergence of a column can be obtained by multiplying the inferred rate by the total branch length of the tree. However, as the alignments contain variable numbers of species, this total branch length represents the contribution of both the rate and the tree topology. To normalize for this effect, the total branch length for each tree fit to an alignment was divided by the total branch length of the reference species tree including only the species in that alignment. The reported substitution rate is therefore a product of this scaling factor and the inferred column rate. The average amino acid or indel substitution rate for a region was calculated as the mean of the respective rate across all columns. Because the indel rates were associated with columns in character alignments, they were mapped back to the original sequence alignment by assigning half of a character's rate to its start and stop positions. Since indel models with limited data were prone to overfitting, rates obtained from character alignments with fewer than five columns were set to zero.

\subsection{Definition and calculation of features}
Features were calculated as in Zarin \textit{et al.} with the following modifications~\cite{Zarin2019}. The regular expression for polar residue fraction was [QNSTCH], which, in contrast to the original study, excludes glycine residues. Additionally, length, expressed in log scale, was replaced with a feature proportional radius of gyration for an excluded-volume polymer~\cite{Flory1949}. Because the radii of gyration of chemically denatured proteins closely match the values expected for equivalent random coils, we felt this feature would better capture the relationship between an IDR's length and its biophysical properties~\cite{Kohn2004}. Finally, several motifs from ELM were replaced with their metazoan counterparts or updated versions of the same entries~\cite{Kumar2021}. These differences are noted in the supplementary data. Furthermore, unlike the previous work, motifs were left as counts and not normalized to the proteome-wide average. Kappa, omega, SCD, hydropathy, PPII propensity, and Wootton-Federhen sequence complexity were calculated with localCIDER 0.1.19~\cite{Holehouse2017}. Isoelectric point was calculated with the Python package isoelectric, which is available on PyPI or at \url{https://isoelectric.org/}~\cite{Kozlowski2016}. Otherwise, features were implemented with custom code. A full list of features and their definitions is given in Table~\ref{stable:features} and Table~\ref{stable:regexes}.

\subsection{Brownian motion and Ornstein-Uhlenbeck analyses}
Brownian motion (BM) model parameters were calculated with two methods. The first used Felsenstein's contrasts algorithm to efficiently calculate roots and contrasts for the disorder scores and features of each region~\cite{Felsenstein1973, Felsenstein1981}. Rates were calculated as the mean of the squares of the contrasts. Though these values are unbiased, they are not maximum likelihood estimates and are inappropriate for use with the Akaike information criterion (AIC)~\cite{Akaike1974}. Thus, they were used for analyses involving only the BM model. For comparison with the Ornstein-Uhlenbeck (OU) model, the BM parameters were calculated by maximizing the likelihood. The OU model parameters were also calculated via maximum likelihood estimation. To avoid issues with model identifiability, the root was treated as a random variable~\cite{Ho2014}. Thus, the covariance matrix, $V$, was parameterized as $V_{ij} = e^{\alpha d_{ij}}$ where $d_{ij}$ is the tree distance between tips $i$ and $j$, and $\alpha$ is the selection strength~\cite{An2008}.

The AICs were calculated for the models of each feature using their maximized likelihoods and two and three parameters for the BM and OU models, respectively. The pairwise differences in the AICs yielded a vector with 82 components, each representing the relative goodness of fit of the OU model over the BM model after accounting for their difference in complexity. The vectors were clustered using the correlation distance metric and the UPGMA algorithm. Clusters were manually chosen for subsequent GO analyses. To enrich these clusters for regions with a high likelihood of feature conservation despite high levels of sequence divergence, this analysis was restricted to regions whose amino acid and indel substitution rates exceeded 1 and 0.1, respectively (Fig.~\ref{sfig:hierarchy_histogram}).

\subsection{GO term analyses}
The 2022-03-22 go-basic release of the Gene Ontology was obtained from the GO Consortium website~\cite{Ashburner2000, 2020GOConsortium}. The gene association file for the 2022\_02 release of the \textit{D. melanogaster} genome annotation was obtained from FlyBase~\cite{Gramates2022}. Obsolete annotations were dropped, and the remaining annotations were filtered by qualifiers and evidence code. The allowed qualifiers were ``enables,'' ``contributes\_to,'' ``involved\_in,'' ``located\_in,'' ``part\_of,'' and ``is\_active\_in.'' The allowed evidence codes were all experimental sources, traceable author statement (TAS), and inferred by curator (IC). The annotations were propagated up the ontology graph and joined with the region sets, so every annotation associated with a gene was associated with the regions derived from that gene. \textit{P}-values were calculated with exact hypergeometric probabilities with regions considered as the sampling unit. For the disorder score analysis, the reference set was the filtered regions, and the enrichment set was regions in the upper decile of the score rate distribution (Fig.~\ref{sfig:scores_histogram}). For the cluster analysis, the reference set was the filtered regions after the additional filtering by substitution rates, and the enrichments sets were the regions in each cluster.

\subsection{Code and data availability}
\begin{sloppypar}
The code used to produce the results and analyses is available at \url{https://github.com/marcsingleton/IDR_evolution2023}. The following Python libraries were used: matplotlib, NumPy, pandas, SciPy, and scikit-learn~\cite{Hunter2007, Harris2020, McKinney2010, Virtanen2020, Pedregosa2011}. Relevant output files are available in the supporting information. There is no primary data associated with this manuscript. All primary data are available from publicly accessible sources described in their corresponding sections.
\end{sloppypar}