\begin{abstract}
\noindent
Here's where my abstract will go!
\end{abstract}

\section{Introduction}
Here's where my introduction will go!

\section{Results}
\begin{figure}[h!]
\includegraphics[width=\textwidth]{scores/out/scores_07E3.png}
\centering
\caption{\textbf{Analyses of disorder scores.}
\textbf{(A)} Example region in the alignment of the sequences in orthologous group 07E3 with their corresponding disorder scores. Higher scores indicate a higher probability of disorder. Disorder score traces are colored by the position of its associated species on the phylogenetic tree. \textbf{(B)} Correlations between disorder scores and feature contrasts in regions. Asterisks indicate statistically significant correlations as computed by permutation tests ($p < 0.001$). \textbf{(C-D)} Example scatter plots showing correlations plotted in panel B. \textbf{(E)} GO term analysis of regions with rapidly evolving disorder scores. Only terms whose $p$-values are below 0.001 are shown.}
\label{fig:scores}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=\textwidth]{substitution/out/substitution.png}
\centering
\caption{\textbf{Amino acid substitution models fit to disorder and order regions.}
\textbf{(A)} Amino acid frequencies of substitution models. Amino acid symbols are ordered by their enrichment in disorder regions, calculated as the disorder-to-order ratio of their frequencies. Error bars represent standard deviations over models fit to different meta-alignments ($n = 25$). \textbf{(B-C)} Exchangeability coefficients of disorder and order regions, respectively, averaged over meta-alignments. \textbf{(D)} $\log_{10}$ disorder-to-order ratios of exchangeability coefficients. \textbf{(E-F)} Rate coefficients of disorder and order regions, respectively, averaged over meta-alignments. The vertical and horizontal axes indicate the initial and target amino acids, respectively. \textbf{(G)} $\log_{10}$ disorder-to-order ratios of rate coefficients.}
\label{fig:substitution}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=\textwidth]{brownian/out/root.png}
\centering
\caption{\textbf{PCA of disorder regions' feature roots.}
\textbf{(A)} The first two PCs of the disorder regions' feature root distributions. The explained variance percentage of each component is indicated in parentheses in the axis labels. \textbf{(B)} The same plot as panel A with the projections of original features onto the components shown as arrows. Only the 16 features with the largest projections are shown. Scaling of the arrows is arbitrary.}
\label{fig:root}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=\textwidth]{brownian/out/rate.png}
\centering
\caption{\textbf{PCA of disorder regions' feature rates.}
\textbf{(A)} The first two PCs of the disorder regions' feature rate distributions. The explained variance percentage of each component is indicated in parentheses. \textbf{(B)} Scatter plot of the disorder regions' feature rates along the first PC against the sum of the average amino acid and indel substitution rates in those regions. \textbf{(C)} The second and third PCs of the disorder regions' feature rate distributions. The explained variance percentage of each component is indicated in parentheses in the axis labels. \textbf{(D)} The same plot as panel C with the projections of original rates onto the components shown as arrows. Only the 16 features with the largest projections are shown. Scaling of the arrows is arbitrary. \textbf{(E-F)} Example alignments of disorder regions from the orthologous groups 0A8A, 3139, 04B0, respectively. The colored bars on the left indicate the hexbin containing that region in panel C.}
\label{fig:rate}
\end{figure}

\begin{figure}[h!]
\includegraphics[width=\textwidth]{hierarchy/out/hierarchy.png}
\centering
\caption{\textbf{Hierarchical clustering of evolutionary signatures.}
The AIC difference between the BM and OU models is measure of their relative goodness of fit to the data with a penalty for the number of parameters in each. Larger values indicate a better fit by the OU model.}
\label{fig:hierarchy}
\end{figure}

\section{Discussion}
Here's where my discussion will go!

\section{Materials and methods}

\subsection{Alignment and species tree provenance}
Alignments of 8,566 single copy orthologs and the corresponding outputs of the missing data phylo-HMM were obtained from the analyses conducted in chapter~\ref{chapter:1}. Likewise, the LG consensus tree generated by the ``non-invariant, 100\% redundancy'' sampling strategy was used as the input or reference where indicated in subsequent phylogenetic analyses.

\subsection{IDR prediction and filtering}
AUCPreD was used to predict the disorder score of each residue of each sequence in the alignments after the gap symbols were removed~\cite{Wang2016}. (Alignments 0204 and 35C2 contained sequences which exceeded the 10,000-character limit and were excluded from subsequent analyses.) The resulting scores were then aligned using the original alignment. The average score for each position was calculated using Gaussian process sequencing weighting over the LG consensus tree~\cite{Altschul1989}. Any positions inferred as ``missing'' by the missing data phylo-HMM or to the left or right of the first or last non-gap symbol, respectively, were excluded. For simplicity, the Gaussian process weights were not re-calculated from a tree pruned of the corresponding tips, and instead the weights corresponding to the remaining sequences were re-normalized. The scores at any remaining positions with gap symbols were inferred by linear interpolation from the nearest scored position.

The average disorder scores were converted into contiguous regions with the following method. Two binary masks were defined as positions where the average score exceeded high and low cutoffs of 0.6 and 0.4, respectively. The low-cutoff mask was subjected to an additional binary dilation with a structuring element of size three to merge any contiguous regions separated by a small number of positions with scores below the cutoff. ``Seed'' regions were then defined as 10 or more contiguous ``true'' positions in the high-cutoff mask, and ``disorder'' regions were obtained by expanding the seeds to the left and right until the first ``false'' position in the low-cutoff mask or the end of the alignment. ``Order'' regions were taken as the complement of the disorder regions in each alignment.

The regions were filtered with the following criteria. First, segments with non-standard amino acid symbols, which overlapped with any position labeled as ``missing'' by the missing data phylo-HMM, or whose number of non-gap symbols was below a length cutoff of 30 residues were removed. Regions whose remaining segments failed the set of phylogenetic diversity criteria detailed in Table~\ref{stable:diversity_criteria} were excluded. The final set contained 11,445 and 14,927 disorder and order regions, respectively, from 8,466 distinct alignments.

\subsection{Fitting substitution models and trees}
To fit amino acid substitution matrices to disorder and order regions, 25 meta-alignments were constructed by randomly sampling 100,000 columns from the respective subsets in the filtered regions. To determine the effect of gaps, the maximum fraction of gaps was set at 0, 50, 100\%. The combination of the region types and sampling strategy yielded six different sets of meta-alignments. A GTR20 substitution model with four FreeRate categories and optimized state frequencies was fit to each meta-alignment using IQ-TREE 1.6.12~\cite{Nguyen2014}. Exchangeability and rate coefficients were normalized, so the average rate of each model was equal to 1. Because exchangeability and rate coefficients are highly correlated across meta-alignments of the same region type, all figures are derived from the maximum 50\% gap fraction meta-alignment sets unless otherwise noted (Fig.~\ref{sfig:heatmap_ematrix}-\ref{sfig:heatmap_corr}).

To obtain estimates of the average substitution rates in each region, separate amino acid and indel models were fit to each alignment. For the amino acid substitution models, the columns in the alignments were manually segregated into disorder and order partitions using the regions derived from the AUCPreD scores. However, to prevent poor fits from a lack of data, a partition was created only if it contained a minimum of 20 sequences with at least 30 non-gap symbols. If one partition met these conditions but the other did not, the disallowed partition was consolidated into the allowed one. If neither partition passed, the alignment was skipped. These rules ensured that the regions represented in the filtered set were fit with substitution models which were concordant with their predicted disorder states. Trees were fit to each partition using an invariant and four discrete gamma rate categories using IQ-TREE 1.6.12~\cite{Yang1994}. The disorder partition used a substitution model derived from the average of the state frequencies and rate coefficients fit to the 50\% gap fraction meta-alignment sets sampled from the disorder regions. The order partition used the LG substitution model~\cite{Le2008}. To prevent overfitting of branch lengths, the trees were restricted to scaled versions of the reference species tree using the -{}-blscale option.

As inference with models that allow insertions and deletions of arbitrary lengths is computationally intractable, a more heuristic approach was taken to quantify the amount of evolutionary divergence resulting from indels in the alignments. For a given alignment, all contiguous subsequences of gap symbols with unique start or stop positions in any sequence were defined as binary characters. Then for each character, a sequence was coded with the symbol 1 if it contained that character or that character was nested in another contiguous subsequence of gap symbols in the sequence and the symbol 0 otherwise. GTR2 models with optimized state frequencies and ascertainment bias correction were fit to the resulting character alignments. A discrete gamma rate category was added for every five character columns, up to a maximum of four. To prevent overfitting of branch lengths, the trees were restricted to scaled versions of the reference species tree using the -{}-blscale option.

Because the rate and branch lengths of a phylogenetic substitution model always appear as products in the likelihood expression, they are not jointly identifiable parameters. Instead, the rate is conventionally taken as equal to one (with inverse count units), and the branch lengths are expressed in terms of the expected number of substitution events. For models with multiple rate categories, the equivalent condition is that the mean of the prior distribution over the rate categories is equal to one. This effectively makes each rate category a scaling factor of the branch lengths. The inferred rate of a column, calculated as the mean of the posterior distribution over the rate categories, is therefore relative to the average across all columns in the alignment. An absolute measure of the evolutionary divergence of a column can be obtained by multiplying the inferred rate by the total branch length of the tree. However, as the alignments contain variable numbers of species, this total branch length represents the contribution of both the rate and the tree topology. To normalize for this effect, the total branch length for each tree fit to an alignment was divided by the total branch length of the reference species tree including only the species in that alignment. The reported substitution rate is therefore a product of this scaling factor and the inferred column rate. The average amino acid or indel substitution rate for a region was calculated as the mean of the respective rate across all columns.

\subsection{Definition and calculation of features}
Features were calculated as in Zarin \textit{et al.} with the following modifications~\cite{Zarin2019}. The regular expression for polar residue fraction was [QNSTCH], which, in contrast to the original study, excludes glycine residues. Additionally, length, expressed in log scale, was replaced with a feature proportional radius of gyration for an excluded-volume polymer~\cite{Flory1949}. Because the radii of gyration of chemically denatured proteins closely match the values expected for equivalent random coils, we felt this feature would better capture the relationship between an IDR's length and its biophysical properties~\cite{Kohn2004}. Finally, several motifs from ELM were replaced with their metazoan counterparts or updated versions of the same entries~\cite{Kumar2021}. These differences are noted in the supplementary data. Furthermore, unlike the previous work, motifs were left as counts and not normalized to the proteome-wide average. Kappa, omega, SCD, hydropathy, PPII propensity, and Wootton-Federhen sequence complexity were calculated with localCIDER 0.1.19~\cite{Holehouse2017}. Isoelectric point was calculated with the Python package isoelectric, which is available on PyPI or at \url{https://isoelectric.org/}~\cite{Kozlowski2016}. Otherwise, features were implemented with custom code. A full list of features and their definitions is given in Table~\ref{stable:features} and Table~\ref{stable:regexes}.

\subsection{Brownian motion and Ornstein-Uhlenbeck analyses}
Brownian motion (BM) model parameters were calculated with two methods. The first used Felsenstein's contrasts algorithm to efficiently calculate roots and contrasts for the disorder scores and features of each region~\cite{Felsenstein1973, Felsenstein1981}. Rates were calculated as the mean of the squares of the contrasts. Though these values are unbiased, they are not maximum likelihood estimates and are inappropriate for use with the Akaike information criterion (AIC)~\cite{Akaike1974}. Thus, they were used for analyses involving only the BM model. For comparison with the Ornstein-Uhlenbeck (OU) model, the BM parameters were calculated by maximizing the likelihood. The OU model parameters were also calculated via maximum likelihood estimation. To avoid issues with model identifiability, the root was treated as a random variable~\cite{Ho2014}. Thus, the covariance matrix, $V$, was parameterized as $V_{ij} = e^{\alpha d_{ij}}$ where $d_{ij}$ is the tree distance between tips $i$ and $j$, and $\alpha$ is the selection strength~\cite{An2008}.

The AICs were calculated for the models of each feature using their maximized likelihoods and two and three parameters for the BM and OU models, respectively. The pairwise differences in the AICs yielded a vector with 82 components, each representing the relative goodness of fit of the OU model over the BM model after accounting for their difference in complexity. The vectors were clustered using the correlation distance metric and the UPGMA algorithm. Clusters were manually chosen for subsequent GO analyses.

\subsection{GO term analyses}

\subsection{Code and data availability}
\begin{sloppypar}
The code used to produce the results and analyses is available at \url{https://github.com/marcsingleton/IDR_evolution2023}. The following Python libraries were used: matplotlib, NumPy, pandas, SciPy, and scikit-learn~\cite{Hunter2007, Harris2020, McKinney2010, Virtanen2020, Pedregosa2011}. Relevant output files are available in the supporting information. There is no primary data associated with this manuscript. All primary data are available from publicly accessible sources described in their corresponding sections.
\end{sloppypar}