\documentclass[11pt,letterpaper,oneside]{book}

% PACKAGE AND ENVIRONMENT CONFIGURATIONS
\usepackage{amsmath}
\DeclareMathOperator*{\argmax}{arg\!\max}
\DeclareMathOperator*{\argmin}{arg\!\min}

\usepackage{graphicx}

% Code listings
\usepackage[outputdir=out]{minted}
\usepackage{tcolorbox}
\tcbuselibrary{breakable}
\tcbset{colframe=white}
\usepackage{lmodern}

% Margins and spacing
\usepackage[margin=1in]{geometry}
\usepackage{lscape}
%\usepackage{setspace}  % double spacing
%\doublespacing
\usepackage[skip=10pt]{parskip}  % Remove indents
\pagestyle{plain}

% Chapter and figure formatting
\usepackage{longtable}
\usepackage{titlesec}  % Modify chapter title format
\titleformat{\chapter}[display]% Shape (chapter label in separate paragraph)
                      {\bfseries\Large}% Format
                      {\MakeUppercase{\chaptertitlename} \thechapter}% Label
                      {0in}% Sep
                      {}% Before code
                      []% After code
\setcounter{tocdepth}{1}  % Only number divisions to section
\usepackage[font=small, labelfont=bf, labelsep=period]{caption}
\renewcommand{\thefigure}{\thechapter.\arabic{figure}}

% References
\usepackage[backend=biber, style=nature, doi=false, url=false]{biblatex}
\addbibresource{refs.bib}
\usepackage{hyperref}

% ENVIRONMENTS
\newenvironment{abstract}
{% Begin code
 \small
 \begin{center}
 \textbf{Abstract}
 \end{center}
 \quotation
}
{% End code
}

\definecolor{incolor}{RGB}{51,63,153}
\definecolor{outcolor}{RGB}{200,78,41}
\definecolor{bgcolor}{RGB}{247,247,247}

\newlength{\promptwidth}
\setlength{\promptwidth}{0.375in}

\newcommand{\prompt}[2]
{\makebox[0pt][r]{\texttt{\color{#2}#1:\ }}\vspace{-\baselineskip}}

\newenvironment
{NotebookIn}
{%
 \VerbatimEnvironment%
 \begin{tcolorbox}[sharp corners,
                   frame empty,
                   breakable,
                   size=fbox,
                   colback=bgcolor,
                   left skip=\promptwidth,
                   title=\prompt{In}{incolor},
                   attach title to upper]%
 \begin{minted}{python}%
}
{%
 \end{minted}%
 \end{tcolorbox}%
}

\newenvironment
{NotebookOut}
{%
 \VerbatimEnvironment%
 \begin{tcolorbox}[sharp corners,
                   frame empty,
                   breakable,
                   size=fbox,
                   colback=white,
                   left skip=\promptwidth,
                   title=\prompt{Out}{outcolor},
                   attach title to upper]%
 \begin{minted}{text}%
}
{%
 \end{minted}%
 \end{tcolorbox}%
}

\newenvironment
{NotebookImage}
{%
 \begin{tcolorbox}[sharp corners,
                   frame empty,
                   size=fbox,
                   colback=white,
                   left skip=\promptwidth,
                   title=\prompt{Out}{outcolor},
                   attach title to upper]%
 \\%
}
{%
 \end{tcolorbox}%
}

% DOCUMENT
\begin{document}

\begin{titlepage}
\begin{center}
    Finding structure in disorder: Evolutionary analyses of disordered proteins in \textit{Drosophila}\\
    \bigskip
    By\\
    Marc Singleton\\
    \vfill
    A dissertation submitted in partial satisfaction of the\\
    requirements for the degree of\\
    Doctor of Philosophy\\
    in\\
    Biophysics\\
    in the\\
    Graduate Division\\
    of the\\
    University of California, Berkeley\\
    \vfill
    Committee in charge:\\
    Professor Mike Eisen, Chair\\
    Professor Liana Lareau\\
    Professor Susan Marqusee\\
    Professor Priya Moorjani\\
    \bigskip
    Spring 2023\\
\end{center}
\end{titlepage}

\thispagestyle{empty}
\null  % Necessary to have text or command before \vfill for proper spacing
\vfill
\begin{center}
    \copyright \  2023 Marc Singleton
\end{center}
\clearpage

\setcounter{page}{1}  % Reset page counter after copyright
\begin{center}
    Abstract\\
    \bigskip
    Finding structure in disorder: Evolutionary analyses of disordered proteins in \textit{Drosophila}\\
    by\\
    Marc Singleton\\
    \bigskip
    Doctor of Philosophy in Biophysics\\
    University of California, Berkeley\\
    Professor Mike Eisen, Chair\\
    \bigskip
\end{center}

Living systems are governed by the interactions between large collections of atoms called macromolecules. The most important class of these macromolecules are proteins, which are the molecular machines that carry out a cell's processes. Though proteins are linear chains of simpler building blocks called amino acids, many proteins accomplish their functions by folding into well-defined three-dimensional structures. For many years scientists believed fixed structures were necessary for protein function, but by the early 2000s evidence had accumulated that segments with no fixed spatial relationship between their atoms are ubiquitous in proteins. Furthermore, because these intrinsically disordered regions (IDRs) are highly flexible and can therefore interact with diverse binding partners, they are essential for many cellular processes related to signaling and regulation.

Although our understanding of the structure and function of IDRs has grown significantly over the past two decades, predicting their functions from their sequences of amino acids remains a significant challenge. Because IDRs are structurally unconstrained, their sequences evolve rapidly and are therefore not amenable to traditional bioinformatics techniques which depend on the precise order of amino acids to make comparisons with known proteins. There is increasing evidence, though, that IDRs conserve distributed features such as their chemical composition or net charge, and a recent study clustered IDRs with similar patterns of conserved features into groups with distinct functions. This study, however, was restricted to IDRs in a set of yeast genomes, so it is unclear if these global relationships between conserved features and function are unique to yeast or a general property of IDR evolution. Thus, in this work I conduct a series of evolutionary analyses of IDRs in the genomes of 33 different species of fruit flies to detect patterns of conservation.

These comparisons, however, require the identification of IDRs with common ancestry which perform equivalent functions across many distinct organisms. Since the first genomes were sequenced in the late 1990s, researchers have developed techniques for identifying and aligning such proteins, called orthologs. While these methods are generally effective, they are conducted by automated computational pipelines and prone to errors when processing the highly divergent sequences that characterize many IDRs. The evolutionary relationships between the genomes of closely related species generally make such mistakes easier to identify, and fortunately over the past five years advances in DNA sequencing technology have yielded dramatic increases in the number of sequenced genomes in the \textit{Drosophila} genus. However, because the existing methods for ortholog identification were designed for fewer or more distantly related genomes, they do not fully leverage such genomic redundancy to minimize errors.

Thus, in the first chapter I develop a novel method for identifying orthologs which addresses this shortcoming and apply it to 33 \textit{Drosophila} genomes to generate a set of aligned orthologs. In the second chapter I then identify rapidly evolving IDRs in these alignments and analyse them with a variety of evolutionary models dissect the forces driving their evolution and detect patterns of conservation. Finally, in the third chapter I discuss several software tools and tutorials for fitting statistical models to data, which were created while pursuing the previous aims.

\clearpage
\setcounter{page}{1}  % Reset page counter after abstract
\pagenumbering{roman}  % Set numbering style to roman for pre-main content

\null  % Necessary to have text or command before \vfill for proper spacing
\vfill
\begin{center}
    To my parents, who were always willing to indulge\\
    another ``experiment'' in their kitchen sink
\end{center}
\vfill
\clearpage

\begin{center}
    \bfseries\Large Acknowledgements
\end{center}
\bigskip
If it takes a village to raise a child, then it must take at least three to raise a scientist. Over the years, I've had countless friends, teachers, and mentors. These relationships have shaped the person I am today, so I'm grateful to all the people who took time out of their busy lives to help me in ways both small and large. However, I want to single out the following people who had an especially important impact on my time in graduate school.\\\\
Thank you to my advisor, Mike, for supporting me as a scientist and a person in equal measure.\\\\
Thank you to my committee, Susan, Liana, and Priya, for giving me the pep talk I needed every year.\\\\
Thank you to my lab, Holli, Stadler, Ciera, Ashley, Victoria, Augusto, and Xiao-Yong, for creating a space where it's okay to ask questions. A special thank you to my bay mates, Jenna and Colleen, for getting me through the hard days with sound advice and a healthy dose of humor.\\\\
Thank you to my undergraduate research mentors, Cathy and Jessica, for teaching me the most exciting science is at the intersection of different fields.\\\\
Thank you to my parents, Amy and David, for everything.\\\\
Thank you to my brothers, Andrew and Daniel, for giving me thick skin in the way only big brothers can.\\\\
Thank you to all the friends I've made in California and the Bay Area, for the much-needed life outside of work.\\\\
Finally, thank you to my partner, Joseph, for going on this crazy ride with me. I couldn't have done this without you.
\clearpage

\tableofcontents
\clearpage  % End page so numbering is set properly
\setcounter{chapter}{-1}  % Set chapter counter to 0
\setcounter{page}{1}  % Reset page counter after pre-main content
\pagenumbering{arabic}  % Set numbering style to arabic for main content

\chapter{Introduction}
\label{chapter:0}
\graphicspath{{chapter0/figures/}}
\input{chapter0/chapter0.tex}

\chapter{Leveraging genomic redundancy to improve inference and alignment of orthologous proteins}
\label{chapter:1}
\graphicspath{{chapter1/figures/}}
\input{chapter1/chapter1.tex}

\chapter{Evolutionary analyses of IDRs reveal patterns of conserved features}
\label{chapter:2}
\graphicspath{{chapter2/figures/}}
\input{chapter2/chapter2.tex}

\chapter{Tools and tutorials for fitting mixture models and HMMs}
\label{chapter:3}
\graphicspath{{chapter3/}}
\input{chapter3/introduction.tex}
\input{chapter3/training_tutorial}

\printbibliography[heading=bibintoc, title=References]  % bibintoc == put in table of contents

\appendix
\setcounter{figure}{0}
\renewcommand{\thefigure}{\thechapter\arabic{figure}}
\setcounter{table}{0}
\renewcommand{\thetable}{\thechapter\arabic{table}}

\chapter{Supporting information for chapter 1}
\label{appendix:a}
\graphicspath{{chapter1/figures/}}
\input{chapter1/chapter1_supporting.tex}

\chapter{Supporting information for chapter 2}
\label{appendix:b}
\graphicspath{{chapter2/figures/}}
\input{chapter2/chapter2_supporting.tex}

\chapter{Supporting information for chapter 3}
\label{appendix:c}
\setlength{\promptwidth}{0in}  % No indent for appendix code
\input{chapter3/training_tutorial_supporting.tex}

\end{document}